src/app/qgslayertreeviewfilterindicator.cpp

```diff
@@ -0,0 +1,207 @@
+/***************************************************************************
+  qgslayertreeviewfilterindicator.cpp
+  --------------------------------------
+  Date                 : Januray 2018
+  Copyright            : (C) 2018 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgslayertreeviewfilterindicator.h"
+
+#include "qgslayertree.h"
+#include "qgslayertreemodel.h"
+#include "qgslayertreeview.h"
+#include "qgsquerybuilder.h"
+#include "qgsvectorlayer.h"
+
+
+QgsLayerTreeViewFilterIndicatorManager::QgsLayerTreeViewFilterIndicatorManager( QgsLayerTreeView *view )
+  : QObject( view )
+  , mLayerTreeView( view )
+{
+  mIcon = QgsApplication::getThemeIcon( QStringLiteral( "/mActionFilter2.svg" ) );
+
+  QgsLayerTree *tree = mLayerTreeView->layerTreeModel()->rootGroup();
+  onAddedChildren( tree, 0, tree->children().count() - 1 );
+
+  connect( tree, &QgsLayerTree::addedChildren, this, &QgsLayerTreeViewFilterIndicatorManager::onAddedChildren );
+  connect( tree, &QgsLayerTree::willRemoveChildren, this, &QgsLayerTreeViewFilterIndicatorManager::onWillRemoveChildren );
+}
+
+
+void QgsLayerTreeViewFilterIndicatorManager::onAddedChildren( QgsLayerTreeNode *node, int indexFrom, int indexTo )
+{
+  // recursively connect to providers' dataChanged() signal
```

Comment:
It looks like there's a significant amount of logic in this class that other indicator managers will need to handle. Could we move some of this to a base class to avoid requiring others to reimplement this?

Reply:
I thought about creating a class that would be sending "high-level" notifications from layer tree, but I do not know yet how it should look like. Probably something that is not useful just for indicators, but also to other pieces of code. E.g. a "watcher" object that sends signals layersAdded, layersWillBeRemoved, groupsAdded, groupsWillBeRemoved. But then there are situations where a layer node is added, but layer is not loaded yet - some users may want to get those signals emitted only when layer is fully loaded, others would not care. Or moving items with drag'n'drop - when a layers/groups can be temporarily located in a tree twice...

So, it would be nice to have the logic for these things wrapped somewhere, but I would like to have something well thought out before adding it to API, so if you don't mind I would keep this bit as-is for now...

Reply:
Fair enough!